File: zertz/marble.py
from pyp5js import *

__COLOURS__ = {
    "C": [18, 151, 147],
    "P": [255, 65, 134],
    "Y": [255, 204, 102]
}

class Marble:
    def __init__(self,
                 x:int,
                 y:int,
                 radius:int,
                 colour:str):
        self.init_x = x
        self.init_y = y
        self.x = x
        self.y = y

        self.colour =__COLOURS__[colour]
        self.on_ring = False
        self.marble_radius = radius
        self.is_clicked = False

    # def marble_params(self                 x:int,
    #              y:int,
    #              radius:int,
    #              colour:str):
    #     self.x = x
    #     self.y = y
    #     self.colour =__COLOURS__[colour]
    #     self.on_ring = False
    #     self.marble_radius = radius

    def _clicked(self):
        d = dist(mouseX, mouseY, self.x, self.y)

        if d < self.marble_radius /2:
            self.xoffset = mouseX - self.x
            self.yoffset = mouseY - self.y
            self.is_clicked = True

        else:
            self.xoffset = 0
            self.yoffset = 0
            self.is_clicked = False

    def _dragged(self):
        if self.is_clicked:
            self.x = mouseX - self.xoffset
            self.y = mouseY - self.yoffset

    def reset_init_pos(self):
        self.x = self.init_x
        self.y = self.init_y

    def set_pos(self, x, y):
        self.x = x
        self.y = y

        self.init_x = x
        self.init_y = y

    def render(self):
        noStroke()
        fill(self.colour)
        circle(self.x, self.y, self.marble_radius)
        fill(255)
        circle(self.x -6, self.y -6, 10)

File: Zertz/Zertz/marble.js
class Marble extends Token {
    constructor(posx,
        posy,
        colour) {
        super();
        // Save the initial position of the marble to reset
        this.init_posx = posx;
        this.init_posy = posy;

        this.posx = posx;
        this.posy = posy;

        this.radius = MARBLE_RADIUS

        if (!colour in MARBLE_COLOURS) {
            console.error("Undefined colour given to marble.");
        }
        this.colour = MARBLE_COLOURS[colour];
        this.on_ring = false;
        this.ring_id = null;
        this.is_clicked = false;
        this.is_hover = false;

        // offset to track while being dragged
        this.xoffset = 0;
        this.yoffset = 0;

    }

    _set_pos(posx, posy) {
        this.posx = posx;
        this.posy = posy;

        this.init_posx = posx;
        this.init_posy = posy;
    }

    _reset_init_pos() {
        this.posx = this.init_posx;
        this.posy = this.init_posy;

    }

    _clicked() {
        let d = dist(mouseX, mouseY, this.posx, this.posy);

        if (d < this.radius / 2) {
            this.xoffset = mouseX - this.posx;
            this.xoffset = mouseY - this.posy;

            this.is_clicked = true;

        } else {
            this.xoffset = 0;
            this.yoffset = 0;
            this.is_clicked = false;

        }
        return this.is_clicked;
    }

    _dragged() {
        if (this.is_clicked) {
            this.posx = mouseX - this.xoffset;
            this.posy = mouseY - this.yoffset;
        }

    }

    _hovered() {
        let d = dist(mouseX, mouseY, this.posx, this.posy);

        if (d < this.radius / 2) {
            this.radius = MARBLE_RADIUS + 4;
            this.is_hover = true;


        } else {
            this.radius = MARBLE_RADIUS;
            this.is_hover = false;
        }
        return this.is_hover;

    }

    _render() {
        noStroke();
        fill(this.colour);
        circle(this.posx, this.posy, this.radius);
        fill(255);
        circle(this.posx - 6, this.posy - 6, 10);

    }

};